### DISCUSSION TITLE
Listen for bootstrap events?

### DISCUSSION THREAD

**Discussion by quasipickle** [alpinejs/alpine#1885] [1. Help]
I'm wondering if v3 has an ability to listen for bootstrap custom events (ex: "shown.bs.modal").  I guess the "." conflicts with Alpine syntax for adding modifiers to events, but is there alternate syntax one can use?  Something like:

```html
<div class = "modal" @[shown.bs.modal]="doSomething()">
```
???

**Comment by KevinBatdorf** [comment]
I've seen this asked a few times. Probably worth making a PR and asking for comment. It should be relatively simple to set up. You just need to be able to preserve known modifiers and then transform the rest similar to  how camel works.

https://github.com/alpinejs/alpine/blob/b46c41fa240cd8af2dcaa29fb60fb1db0389c95a/packages/alpinejs/src/utils/on.js#L13

Maybe `<div @shown.bs.modal.dot>` which just transforms everything before `.dot` so you can preserve `shown.bs.modal.dot.window` etc, and keep it framework agnostic

**Reply by quasipickle** [reply]
I could give it a shot.  Is there a way to run this project from the source-code _without_ NPM?  My hosting provide doesn't have it so I'm getting an error when trying to import `@vue/reactivity`

**Reply by KevinBatdorf** [reply]
I think can just clone the repo, run `npm ci` then `npm run watch` right?

**Reply by quasipickle** [reply]
I don't have `npm` on my server and I can't install it.  Is that a non-starter then?

**Reply by KevinBatdorf** [reply]
Yeah I guess so, `yarn` maybe? I don't know if you can do the same or not.

@ryangjchandler - didn't you add the camel case stuff? Was this dot notation ever discussed? Might be useful if it suddenly adds support for interacting with bootstrap.

**Reply by SimoTod** [reply]
@quasipickle you don't need npm on your server, you run those commands locally to generate the updated version in the dist folder (then you can use on your server).

To be fair, you don't even need a server, you can run a dev server locally.

I think the syntax need some discussion, i find `.dot` weird, camel actually modifies the string, we are using dot just as a separator which is going to be a nightmare to implement: the parsing is the same for all directives.

It would be easier to implement something like x`-on:(my.event).window`

**Reply by KevinBatdorf** [reply]
I figured you could just stop at `dot` (not that I'm fond of the name either though) and use everything before it? Is the array consistent with the ordering? For example, 

```js
const modifiers = ['shown', 'bs', 'modal', 'dot', 'window']
return modifiers.slice(0, modifiers.indexOf('dot')).join('.')
```

**Reply by SimoTod** [reply]
I suspect that it would be too late, the event name shouldn't be in the modifier part but in the value one (https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives.js#L128) otherwise the code will just set a listener for the first word

**Reply by KevinBatdorf** [reply]
Ah that's right, because `shown` wouldn't be a modifier. 

Maybe a custom directive would be the best approach, like`x-bs:shown.bs.modal=""`

**Reply by quasipickle** [reply]
@SimoTod I don't do any dev work on my local computer - it's all remote to servers.

Since I use Alpine in my job, I used work servers–which have npm installed–and came up with this implementation of @KevinBatdorf's code:

```javascript
if (modifiers.includes("dot")) [event, modifiers] = dotSyntax(event, modifiers);
```
... further down...
```javascript
function dotSyntax(event, modifiers) {
    event     += '.' + modifiers.slice(0, modifiers.indexOf('dot')).join('.')
    modifiers = modifiers.slice(modifiers.indexOf('dot')+1);

    return [event, modifiers];
}
```
This manages to capture the `shown.bs.modal` event.

I'm happy to make a PR - but wanted to see if this technique is even worth PR-ing.

**Reply by KevinBatdorf** [reply]
Ah nice. I didn't look close enough to realize you have access to the event name as well. 

I think it's useful. I'm not sure about `dot` though as the name either, although nothing better comes to mind. Maybe dot works.

I think it's worth a PR. Does it work though? I don't think the return signature matches. Looking at `camel` you want to return the event name. Maybe you can mutate the modifiers directly? Or via `wrapHandler`? (I'm just throwing out quick ideas though and not sure if they will work)

**Reply by quasipickle** [reply]
Honestly, I'd like `@(shown.bs.modal).window` syntax better.  But something before on.js was messing up the event name, so this was easier.  Maybe I'll take another look.

I'm not sure what you're referring to by "return signature".  Each modifier does it's own thing.  Yes, `camelCase` returns the event name, but other modifiers do different things.

All the current tests are passing, if that counts for anything.

**Reply by quasipickle** [reply]
...could do a | separator I guess, like `@shown.bs.modal|.window`

**Reply by KevinBatdorf** [reply]
> I'm not sure what you're referring to by "return signature". Each modifier does it's own thing. Yes, camelCase returns the event name, but other modifiers do different things.

You're right, I ignored the first part of your code for some reason. Need another coffee.

**Reply by quasipickle** [reply]
I had a brain wave over lunch: Use the regular `-` syntax in the markup, then convert it to dot syntax with the `.dot` modifier.  That way it behaves just like `.camel` modifier.  I'll get some tests written and make a PR.

**Reply by KevinBatdorf** [reply]
I sort of think that PR went in the wrong direction actually. I personally preferred the previous one because it doesn't require users to use dashes where they would naturally want to use `.` 

Did you look into `@(something.with.dots).window`? In that case does `(something.with.dots)` show as the event name? That would be easier to strip out and essentially be future proof for any other alternative event naming. I'm not sure of issues with having `(` in an HTML attribute name though. Maybe `[]`?

**Reply by quasipickle** [reply]
It's no different than having users use dashes when initially they'd want to use camelCase.  The _reason_ is different, but the end result - a modification of the event name - is the same.  That's a strength of this approach - its behaviour is the same as the existing `.camel` modifier and doesn't introduce new syntax.

I didn't look into using () or [] - this seemed much simpler.  I guess I could still take a look, but I do think the PRs approach is superior.

**Reply by KevinBatdorf** [reply]
Yeah I guess if it matches camel case then it's probably better. You're probably right. Let's see how Caleb responds.

**Reply by SimoTod** [reply]
Doesn't boostrap have some events with both dashes and dots though?

**Reply by SimoTod** [reply]
I like the grouping thing but I'm biased :)
It should be doable changing the regex at https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives.js#L137 to something like `/:([a-zA-Z0-9\-:]+|\([a-zA-Z0-9\-:\.]+\))/` and trimming out the parentheses at https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives.js#L143

Caleb might not like it though this is just my personal opinion

**Reply by quasipickle** [reply]
I was looking into that as you posted the regex.  We'd also have to modify the regex for modifiers too.

I guess we could do both, but yeah - let's see what the Grand Poobah thinks.
